### PR TITLE
Better logic for create new labels based on existing images and settings

### LINKED ISSUE #470: Clicking the "add labels" button when viewing a lazy (zarr/dask) image should create a zarr-backed labels array.
## 🚀 Feature

Currently, add_labels will try to create an in-memory labels array of the same shape as the currently viewed image. This will break if the image layer is a bigger-than-memory dask array. A possible solution is to create an out-of-core dataset backed by zarr.

### PR DESCRIPTION
**Pull request by Czaki**
# References and relevant issues

closes #470

# Description

This PR implements:

1. Selection of new label type in settings (possible extension is to suggest conversion when the user selects too big a label to fit in given type) 
2. Logic to select new labels layer class. It implements three options 
	* **fit in ram** (default) - if data are small, create a numpy array; otherwise, create zarr.Array in memory
	* **Follow image class** - Create a label layer using the same class as the base image
	* **Follow image class but fallback to fit in RAM if needed** - mix of both above, if data are to big, fallback to `zarr.Array`. 

This PR uses only classes of Image layers data to determine new label data class. 

We may need to decide how exactly to support edge cases.  

Waiting on #6216

### CODE REVIEW

**Comment by jni** [comment]
Amazing! Love closing three-digit issues 😂 

I think that we should not create an array of the same type as the input array, but rather, create based on availability of the following libraries, in order:
- tensorstore
- zarr
- dask

(Indeed dask is currently useless, because it doesn't support fancy indexing, so you can't paint into it.)

This is because the libraries have different performance characteristics: tensorstore is much faster to read to than zarr. And with dask, I don't know how writing is implemented, but it seems plausible that it stores the written values uncompressed in memory, which would be terrible for this application.

So, regardless of what the input type is, I think we should use the above types, depending on what the user has installed, and perhaps there should be a warning if the user doesn't have tensorstore, since the experience is much better with it.

**Comment by Czaki** [comment]
> following libraries, in order:
> 
>    * tensorstore
> 
>    * zarr
> 
>    * dask

Selecting `tensorstore` is non default dependency and may be problematic. 

It changes behavior between environments, as they may be plugins that support `zarr` but do not support `tensorstore` (for example writer, but also some others). 

On the other side, I see performance benefits. So I think it should be settable in settings as we do not create bugfix releases really fast.

**Comment by psobolewskiPhD** [comment]
Just going to add that would be nice to close this if possible:
https://github.com/napari/napari/issues/5545

**Comment by Czaki** [comment]
Ok. So few questions.

On meeting, we assume that the default behavior should be follow the class. With exception for dask (not possible to fast paint). 
What should be the logic for dask? Use zarr (#6297)? Try to use tensorstore and then zarr? Ask user to use right click menu for select?

Should I keep "fit in ram" option, or remove it and just assume that the user will use the right click menu? 

cc @jni

**Comment by Czaki** [comment]
> Could you clarify what you mean for those that weren't at the meeting?

If list of image/lables layer has same class, scale, shape (and chunks) create new layer with using same class, scale, shape. 
With exception for dask, as our painting code cannot paint over dask arrays, we need to select another class to store data.

**Comment by jni** [comment]
Ok, so, seeing my comment in #6297, I am now leaning towards *not* assuming that we have zarr, so the button could/should be greyed out with a reason (no zarr installed).

I agreed with your argument @Czaki that it would be bad for the button click to have different behaviour depending on whether tensorstore is installed so let's leave that for right-click.

> Should I keep "fit in ram" option, or remove it and just assume that the user will use the right click menu?

I prefer to leave the RAM-based logic for a separate PR. That will make this one simpler and more self-contained, and limit the number of branches, which will make debugging harder.

> our painting code cannot paint over dask arrays, we need to select another class to store data.

I would rephrase this as dask arrays can't handle our totally-standard painting code. 😂

**Comment by psobolewskiPhD** [comment]
> Ok, so, seeing my comment in #6297, I am now leaning towards _not_ assuming that we have zarr, so the button could/should be greyed out with a reason (no zarr installed).

I think the opposite 🤣 
This awesome PR (it's one of the biggest napari issues that making a labels layer breaks the viewer) and builtins need zarr? what more motivation for including it do we need?

**Comment by jni** [comment]
> it's one of the biggest napari issues that making a labels layer breaks the viewer

Yes, graying out the button solves this problem. ;)

> what more motivation for including it do we need?

again, it's not that I don't like zarr or think it's useful, and I think in *most* installs you'd just have it and things would work. But we don't actually need it for our basic functionality (looking at arrays), and it's actually a reasonably complex dependency (because of numcodecs). There's many communities that (to my knowledge) have yet to make use of zarr (e.g. astro, materials science), so for them it is just dead weight.

**Comment by psobolewskiPhD** [comment]
> Yes, graying out the button solves this problem. ;)

I think we differ in the definition of `solves` 😜

**Comment by DragaDoncila** [comment]
I know I keep flip flopping but I'm kinda with Peter on this one... I feel like making a new labels layer and opening zarrs is so important for napari to be useful that it warrants a zarr dependency...

I would be so annoyed if I couldn't make a new labels layer because I didn't have zarr installed and I hovered over it and it told me to "please install zarr". Is it easy enough to install? Yes. So easy in fact that you should've done it for me 😂. But it would totally break my workflow and be very annoying. I get caught out with not having zarr installed all the time and every time I realize it I'm like "why the heck wasn't it already there".

**Comment by jni** [comment]
> I know I keep flip flopping but I'm kinda with Peter on this one... I feel like making a new labels layer and opening zarrs is so important for napari to be useful that it warrants a zarr dependency...

This is overstating the problem. You can make a new labels layer now with NumPy and that is good enough for (I reckon) 99% of users.

Similarly, opening zarrs is only an issue for bioimaging (and some geo) communities, and even then for only a subset. Even coming from a bioimaging conference, only a subset of folks there had even heard of zarr there. Outside of bio, it's blank stares.

> Is it easy enough to install? Yes.

This is the case currently, on some platforms. The number of maintainers of zarr-python is shockingly low and if something starts not compiling on some platforms for whatever reason, e.g. Python 3.14 changing some binary compatibility stuff, or NumPy 2.0 breaking things, there could be significant periods where installing zarr is hard, and that would suddenly prevent a bunch of people from installing napari — including the 99% of napari users who don't use zarr for anything.

> I get caught out with not having zarr installed all the time and every time I realize it I'm like "why the heck wasn't it already there".

I fully understand this frustration: I get caught out not installing matplotlib into my new envs but I don't think we should depend on it. (Even though it would be useful for a bunch of stuff.)

-------

Anyway, in the interest of moving forward I'd be ok with postponing this discussion (let's see what @Czaki says), but I want to point out that (a) napari-lite needs to happen for a lot of reasons, and (b) it can only be non-hypothetical if we actually start taking steps towards it. Again, I would be fine with uploading two different packages to PyPI, napari-base and napari, with the latter being equivalent to our current `napari[optional]`.

**Comment by psobolewskiPhD** [comment]
Totally forgot about this!
From a point of view of painting, we're limited by GL_MAX_TEXTURE_SIZE too for a labels layer.
I tried a `zarr.zeros((16_000, 16_000), dtype='u2')` (under my GL limit) and added it as labels.
Adding is instant, no memory issue. 🎉 
But...painting was brutally bad 😭 Beach ball 30-60 sec.

**Comment by jni** [comment]
@psobolewskiPhD some qs:

- are you using zarr v2 or v3?
- is that array chunked by default? I expect that you should use chunking (say 1K x 1K) because every chunk that gets touched needs to be compressed and rewritten. This is Fast in general but would be slow for hundreds of MB per second.

**Comment by psobolewskiPhD** [comment]
zarr-python 3.0
For `zarr.zeros((16_000, 16_000), dtype='u2')` I get:
```
In [23]: data.chunks
Out[23]: (16000, 16000)
```
With:
`data = zarr.zeros((16_000, 16_000), dtype='u2', chunks=(1500, 1500))`
painting is much better., but still quite choppy.
OK when zoomed in, chippy when zoomed out -- maybe has to do with chunk boundaries?

**Comment by jni** [comment]
What if you try `compressors=None`?

I'm thinking that zarr default should only be used if ndim > ndisplay, because we need to instantiate the full array anyway if it's a single slice, so there are no actual memory savings.

**Comment by brisvag** [comment]
I looked into fixing merge conflicts for you @Czaki, but this is a bit too hard for me to bring up to date.

Looking at the above discussion, I think in the interest of moving forward to ensure this doesn't stall again would be to:

- implement as suggested by @jni so that if zarr is not installed, the button is greyed out, but otherwise it's magically working (a direct improvement from the current "napari crashes" behaviour)
- open already a followup PR to move zarr from the from the optional deps to the mandatory ones, so we can have that discussion separately

AFAICT there are no other hangups; @Czaki was there something else you need to get this moving?

**Comment by psobolewskiPhD** [comment]
Related: https://github.com/napari/napari/pull/9469
Very large performance improvement for painting in zarr and huge for painting in dask.
Should open up multiscale painting...

**Comment by Czaki** [comment]
I think that it will be better to implement from scratch, but not close this PR before new one is opened.
